### ISSUE TITLE
Produce an ESM compatible file in the distribution (in addition to UMD)

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor.js#801]
With strict mode enabled (see https://github.com/asciidoctor/asciidoctor.js/pull/763) we are now able to produce an ESM compatible file.
For compatibility reason, we should keep the UMD file and produce an additional file:

```
dist
├── browser
│   ├── asciidoctor.js
│   └── asciidoctor.min.js
├── css
│   └── asciidoctor.css
├── esm
│   ├── asciidoctor.js
│   └── asciidoctor.min.js
├── graalvm
│   └── asciidoctor.js
└── node
    └── asciidoctor.js

5 directories, 7 files
```

**Comment by Mogztter** [comment]
We should use the field "module" in the `package.json` file to reference this new ESM compatible file.
https://stackoverflow.com/a/42817320/3600979

To be consistent, we should export the "constructor" function:

```html
<script type="module">
import Asciidoctor from './build/asciidoctor-browser.js';

const asciidoctor = Asciidoctor()
console.log(asciidoctor.convert('Hello *world*'))
// prints:
// <div class="paragraph">
// <p>Hello <strong>world</strong></p>
// </div>
</script>
```

We should also export the `Opal` global for advanced usage.
We might also export a "default" instance and/or common functions.


```html
<script type="module">
import { Asciidoctor, asciidoctor as processor, convert, Opal } from './build/asciidoctor-browser.js';

const asciidoctor = Asciidoctor()
const nil = Opal.nil
console.log(asciidoctor.convert('Hello *world*'))
console.log(processor.convert('Hello *world*'))
console.log(convert('Hello *world*'))
// prints:
// <div class="paragraph">
// <p>Hello <strong>world</strong></p>
// </div>
</script>
```

**Comment by iovuio** [comment]
How can  `./build/asciidoctor-browser.js` be obtained?

**Comment by Mogztter** [comment]
@iovuio You mean from the example above? It was just an example to illustrate the possibilities so this ESM compatible file does not exist yet. Since it's a breaking change it's planned for Asciidoctor.js 3.0.0.

**Comment by oncletom** [comment]
+1 for the `module` field, it would make Asciidoctor compatible with Deno — cf. [its search engine Pika](https://www.pika.dev/search?q=asciidoctor).

**Comment by Mogztter** [comment]
I can produce an ESM that's going to work in the browser but we are not ready yet to produce an ESM for Node/Deno.

My concern is that if I add `module` field then Node.js might try to use this file but it won't work because this file is not designed to work on a Node.js environment... 

As a first step, the field `browser` will point to an ESM and I will remove the UMD from the distribution to prepare the future.

**Comment by Mogztter** [comment]
Still confused about how we should distribute a dedicated file for Node and another one for the browser with `pkg.module`. Do we need to produce a single file and use environment/feature detection? If someone has good resources on the subject I'm more than interested!

**Comment by iovuio** [comment]
@Mogztter crazy thought.
- rename current package to **asciidoctor.js-node**
- deprecate **asciidoctor.js**
- release browser ready version as new package **asciidoctor.js-browser**
- whenever a deno+node ready version will be available release **asciidoctor.js-server**
- whenever a unified version will be available, if ever, revive **asciidoctor.js** from the dead, deprecate all others

**Comment by mojavelinux** [comment]
I'm not keen on that idea. That will only create confusion that will lead to more hours of support. It's best to try to stick with one package (or, if absolutely necessary, introduce a specialized page based on ESM that just stays that way).

**Comment by Mogztter** [comment]
I got confused by the `"type": "module"` used by Node.js to enable ESM and `"module": "asciidoctor.esm.js"` used by web bundlers, pika...

So we might be fine, I will create a dummy package to make sure that everything is working as expected :sweat_smile: 

> crazy thought.
rename current package to asciidoctor.js-node
deprecate asciidoctor.js
release browser ready version as new package asciidoctor.js-browser
whenever a deno+node ready version will be available release asciidoctor.js-server
whenever a unified version will be available, if ever, revive asciidoctor.js from the dead, deprecate all others

> 'm not keen on that idea. That will only create confusion that will lead to more hours of support.

I agree with Dan, we should try to avoid renaming the package at all cost. We already renamed the package once to align with the conventions in the Node/JavaScript ecosystem.

> It's best to try to stick with one package

:+1:

**Comment by Mogztter** [comment]
I've decided to *not* introduce any change to the API and to produce an ECMAScript Module for the browser:

```
dist
├── browser
│   ├── asciidoctor.esm.js
│   ├── asciidoctor.esm.min.js
│   ├── asciidoctor.js
│   └── asciidoctor.min.js
├── css
│   └── asciidoctor.css
├── graalvm
│   └── asciidoctor.js
└── node
    └── asciidoctor.js

4 directories, 7 files
```

I still need to figure out how to build a fully-working version of Asciidoctor.js for Node.js using ESM.

**Comment by Mogztter** [comment]
Fixed in 28d3fad81ff8bf663cb97ae54c9d7997f8563d90.
`dist/browser/asciidoctor.js` is now an ES6 module (it was previously a UMD).
I've decided to publish only an ES6 module. It should still be possible to support browsers that do not support native JavaScript modules using a Web bundler.
